File: 2016/2/2.js
const input = (await Deno.readTextFile("./2.txt")).split("\r\n");

// Part 1

let c = 2, r = 2;
let pw = "";

for (const str of input) {
  for (const chr of str.split("")) {
    switch (chr) {
      case "U": {
        if (r != 1) r--;
        break;
      }
      case "D": {
        if (r != 3) r++;
        break;
      }
      case "L": {
        if (c != 1) c--;
        break;
      }
      case "R": {
        if (c != 3) c++;
        break;
      }
    }
    // console.log(chr, "=> key:", c + ((r - 1) * 3))
  }
  pw += `${c + ((r - 1) * 3)}`;
}

console.log(pw);

// Part 2

let c2 = 0, r2 = 2;
let pw2 = "";
const kp = [
  ["", "", "1", "", ""],
  ["", "2", "3", "4", ""],
  ["5", "6", "7", "8", "9"],
  ["", "A", "B", "C", ""],
  ["", "", "D", "", ""],
];

for (const str of input) {
  for (const chr of str.split("")) {
    switch (chr) {
      case "U": {
        if (r2 != 0 && kp[r2 - 1][c2] != "") r2--;
        break;
      }
      case "D": {
        if (r2 != 4 && kp[r2 + 1][c2] != "") r2++;
        break;
      }
      case "L": {
        if (c2 != 0 && kp[r2][c2 - 1] != "") c2--;
        break;
      }
      case "R": {
        if (c2 != 4 && kp[r2][c2 + 1] != "") c2++;
        break;
      }
    }
    // console.log(chr, "=> key:", kp[r2][c2]);
  }
  pw2 += kp[r2][c2];
}

console.log(pw2);

File: 2025/2/2.jl
file = open("2.txt", "r")
input = read(file, String)
input = split(input, ",")
inputCopy = input
close(file)

ids = Vector{Int128}(undef, 0)

function check(n)
    nStr = string(n)
    nSize = length(nStr)
    if (isodd(nSize))
        return false
    end

    num = trunc(Int128, nSize / 2)
    u = nStr[1:num]
    l = nStr[num+1:end]
    # println(u, "|", l)
    return (u == l)
end

for sub in input
    rng = split(sub, "-")
    from = parse(Int128, rng[1])
    to = parse(Int128, rng[2])
    r = range(from, to)
    for id in r
        if check(id)
            append!(ids, id)
        end
    end
end

println(sum(ids))

# Part 2 coppied and adapted from:
# https://github.com/JoanaBLate/advent-of-code-js/blob/1f6e53f4f9deb050ec238b6156e6ca1e1ecc5154/2025/day02-solve2.js

ids2 = Vector{Int128}(undef, 0)

function checkToken(token, tokenSize, length)
    if (tokenSize == length)
        return false
    end

    if (tokenSize % length != 0)
        return false
    end

    pattern = token[1:length]
    cursor = Int64(1)

    while (true)
        cursor = cursor + length

        # println("cursor: ", cursor, " ; tokenSize: ", tokenSize)

        if (cursor > tokenSize)
            return true
        end

        target = token[cursor:cursor+length-1]

        # println("comparing from \"", token, "\": \"", pattern, "\" and \"", target, "\"")

        if (pattern != target)
            break
        end
    end

    return false
end

for sub in inputCopy
    rng = split(sub, "-")
    from = parse(Int128, rng[1])
    to = parse(Int128, rng[2])
    r = range(from, to)
    for id in r
        str = string(id)
        len = length(str)
        # println("str: ", str, " ; len: ", len)
        if checkToken(str, len, 1)
            append!(ids2, id)
            continue
        end
        if checkToken(str, len, 2)
            append!(ids2, id)
            continue
        end
        if checkToken(str, len, 3)
            append!(ids2, id)
            continue
        end
        if checkToken(str, len, 4)
            append!(ids2, id)
            continue
        end
        if checkToken(str, len, 5)
            append!(ids2, id)
            continue
        end
        if checkToken(str, len, 6)
            append!(ids2, id)
            continue
        end
        if checkToken(str, len, 7)
            append!(ids2, id)
            continue
        end
    end
end

# println(ids2)
println(sum(ids2))
